### PR TITLE
Support qutip-v5

### PR DESCRIPTION
**Pull request by BoxiLi**
This is a replacement of #88.

- Add support of qutip=5.0 to `setup.cfg`
- New GitHub actions are added to test qutip-qip against the dev.major version of qutip
- Fix the importation errors for `qutip.Options`, use `qutip.SolverOptions` instead.
- Use `Qobj.full()` to transfer the matrix to NumPy arrays before using `assert_all_close`, since `Qobj.__array__` is removed.
- Tests are adapted for the new QobjEvo in qutip.
  - Replace the `QobjEvo.ops` and `QobjEvo.cte` as they are removed. Create separate test functions to compare `QobjEvo`s.
  - Remove the use of `QobjEvo.tlist`.
  - Slightly modify the pulse and noise module to correctly handle the continuous/discrete pulse in `QobjEvo`.
- Fix/Improve some other unstable/wrong tests.

### CODE REVIEW

**Comment by BoxiLi** [comment]
@hodgestar @Ericgig I made a version of qutip-qip that is compatible with the current qutip `dev.major` branch. Would you like to have a quick look?

Not to review if the changes are correct for qutip-qip, I can handle that. But rather, if those changes are indeed what we want other users to make to update from version 4 to 5. This could also provide some examples that we can later write in the documentation when we release qutip5.

File: src/qutip_qip/compiler/gatecompiler.py [hunk 1]
```diff
@@ -247,6 +249,18 @@ def _concatenate_pulses(
                 compiled_tlist[pulse_ind].append(execution_time)
                 compiled_coeffs[pulse_ind].append(coeffs)
 
+        final_time = np.max([tlist[-1] for tlist in compiled_tlist])
```
Comment (hodgestar) [inline src/qutip_qip/compiler/gatecompiler.py:252]: It looks like this is unrelated to QuTiP 5. I'm happy for it to stay in this PR potentially, but what does it add / do?

File: src/qutip_qip/device/cavityqed.py [hunk 2]
```diff
@@ -147,7 +147,14 @@ def eliminate_auxillary_modes(self, U):
...
             + [identity(2) for n in range(self.num_qubits)]
         )
-        return psi_proj.dag() * U * psi_proj
+        result = psi_proj.dag() * U * psi_proj
+        # In qutip 5 multiplication of matrices
+        # with dims [[1, 2], [2, 2]] and [[2, 2], [1, 2]]
+        # will give a result of
+        # dims [[1, 2], [1, 2]] instead of [[2], [2]].
+        if result.dims[0][0] == 1:
+            result = result.ptrace(list(range(len(self.dims)))[1:])
```
Comment (hodgestar) [inline src/qutip_qip/device/cavityqed.py:156]: This looks like it will bite other users. Is there some better way to support it? Maybe `psi_proj` dims need to be fixed before the multiplication somehow?

**Review by hodgestar — APPROVED** [review]

File: src/qutip_qip/compiler/gatecompiler.py (hunk 1 above)
Comment (BoxiLi) [reply src/qutip_qip/compiler/gatecompiler.py:252]: This is actually related to the fact that evaluation out of the tlist will give non-zero values in qutip 5. Because of this, I need to determine the final time and add some zeros to the end.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -10,31 +10,37 @@ jobs:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        python-version: [3.6, 3.9]
```
[the remark this answers is not in the document]
Comment (BoxiLi) [reply .github/workflows/test.yml:13]: Ops, will add this back.

File: src/qutip_qip/device/cavityqed.py (hunk 2 above)
Comment (hodgestar) [reply src/qutip_qip/device/cavityqed.py:156]: Since we'd added looking at this to the TODO list, let's leave it for now.

File: .github/workflows/test.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (hodgestar) [reply .github/workflows/test.yml:13]: Once this is added back, I'm happy for the PR to be merged.

File: src/qutip_qip/device/cavityqed.py (hunk 2 above)
Comment (BoxiLi) [reply src/qutip_qip/device/cavityqed.py:156]: @hodgestar, @Ericgig, I was collecting some common error messages that I saw while updating to v5. And I notice this again. This issue still remains. It is still somewhere in a TODO list?
